backend-rust/src/migrations.rs

```diff
@@ -0,0 +1,264 @@
+                let new_version = current.migration_to_next(&pool).await?;
+                info!("Migrated database schema to version {} successfully", new_version.as_i64());
+                current = new_version
+            }
+            Ok::<_, anyhow::Error>(())
+        })
+        .await
+        .transpose()?;
+    Ok(())
+}
+
+/// Check whether the current database schema version matches the latest known
+/// database schema version.
+pub async fn ensure_latest_schema_version(pool: &PgPool) -> anyhow::Result<()> {
+    if !has_migration_table(pool).await? {
+        anyhow::bail!(
+            "Failed to find the database schema version.
+
+To allow running database migrations provide the `--migrate` flag.
+Use `--help` for more information."
+        )
+    }
+    let current = current_schema_version(pool).await?;
+    if current != SchemaVersion::LATEST {
+        anyhow::bail!(
+            "Current database schema version is not the latest
+    Current: {}
+     Latest: {}
+
+To allow running database migrations provide the `--migrate` flag.
+Use `--help` for more information.",
+            current.as_i64(),
+            SchemaVersion::LATEST.as_i64()
+        )
+    }
+    Ok(())
+}
+
+#[derive(Debug, derive_more::Display)]
+#[display("Migration {version}:{description}")]
+struct Migration {
+    /// Version number for the database schema.
+    version:     i64,
+    /// Short description of the point of the migration.
+    description: String,
+    /// Whether the migration does a breaking change to the database schema.
+    /// This can be used for the API to ensure no breaking change has happened
+    /// since the supported schema version.
+    destructive: bool,
+}
+
+impl From<SchemaVersion> for Migration {
+    fn from(value: SchemaVersion) -> Self {
+        Migration {
+            version:     value.as_i64(),
+            description: value.to_string(),
+            destructive: value.is_destructive(),
+        }
+    }
+}
+
+/// Represents database schema versions.
+/// Whenever migrating the database a new variant should be introduced.
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone, Copy, derive_more::Display)]
+#[repr(i64)]
+pub enum SchemaVersion {
+    #[display("0000:Empty database with no tables yet.")]
+    Empty            = 0,
+    #[display(
+        "0001:Initial schema version with tables for blocks, transaction, accounts, contracts, \
+         tokens and more."
+    )]
+    InitialFirstHalf = 1,
+}
+impl SchemaVersion {
+    /// The latest known version of the schema.
+    const LATEST: SchemaVersion = SchemaVersion::InitialFirstHalf;
+
+    /// Parse version number into a database schema version.
+    /// None if the version is unknown.
+    fn from_version(version: i64) -> Option<SchemaVersion> {
+        match version {
+            0 => Some(SchemaVersion::Empty),
+            1 => Some(SchemaVersion::InitialFirstHalf),
+            _ => None,
+        }
+    }
```

Comment:
Implementing the `try_from` trait  might be more common than a custom conversion function.

```suggestion
impl TryFrom<i64> for SchemaVersion {
    type Error = anyhow::Error;

    fn try_from(version: i64) -> Result<Self, Self::Error> {
        match version {
            0 => Ok(SchemaVersion::Empty),
            1 => Ok(SchemaVersion::InitialFirstHalf),
            _ => anyhow::bail!("Unknown database schema version"),
        }
    }
}
```

Reply:
ok, it makes also sense to group them into the implementation trying to minimize the occurance that we `forget` to update the function in the future when introducing new schemaVersions.